[GEOS-11455] Update developers guide with GSIP-220 security policy (#7759)
geoserver/geoserver: doc/en/developer/source/policies/security.rst

.. _security_procedure:

Security Procedure
==================

This page covers some of the lines of communication and tools available to enact our public security policy (see `SECURITY.md <https://github.com/geoserver/geoserver/blob/main/SECURITY.md>`__).

Vulnerability reporting and discussion
--------------------------------------

> Our security policy asks community to take care in reporting and discussion of security vulnerabilities:

Reviewer: ```suggestion
Our security policy asks the community to take care in reporting and discussion of security vulnerabilities:
```

 **Reporting a Vulnerability**
 

>  If you encounter a security vulnerability in GeoServer please take care to report in a responsible fashion:

Reviewer: ```suggestion
 If you encounter a security vulnerability in GeoServer please take care to report it in a responsible fashion:
```

Guidance on communication challenges
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* Issue tracker: If an issue is opened with a security vulnerability, send the link to SECURITY.md to the reporter and close the issue.

* Public Email: If a geoserver-devel or geoserver-user email comes in reporting the vulnerability (or discussing a vulnerability) link to the the SECURITY.md file and invite parties to discussion on geoserver-security email list privately.
  
* Private Email: If you are contacted directly, share the SECURITY.md, and link to commercial support providers (which may include your employer of course).

* Social media: Ask parties to respect coordinated vulnerability disclosure policy.

  Recognize that such communication is outside our control and do not further engage. 
  
* Public CVE: Most reliable response is to locate the CVE on advisory database and issue a pull-request clarification.
  
  * https://github.com/advisories?query=geoserver
  * https://github.com/advisories?query=geotools
  
  You may also go through the steps of "disputing" a CVE, this involves contacting the original numbering authority (each of which have their own procedures).
  

>   Keepin in mind that security researcher may have obligation to report to their national numbering authority.

Reviewer: ```suggestion
  Keep in mind that security researchers may have obligations to report to their national numbering authority.
```

  Ask that they withhold details and work within our SECURITY.md policy (key phrase "coordinated vulnerability disclosure").
  
Kindness:

* This is a long game, so prioritize sustainability and our geoserver-security volunteers.
* Recognize that the reporter is likely stressed, and may not have much flexibility due legal obligations around vulnerability disclosure.
* This is a sustainability challenge, so do not be shy about seeking sponsorship.

geoserver-security list participation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Participation in the geoserver-security, like commit access, is volunteer based and reflects trust.

> * To volunteer email geoserver-security@lists.osgeo.org stating your interest.

Reviewer: ```suggestion
* To volunteer, email geoserver-security@lists.osgeo.org stating your interest.
```

> * If you are unable to participate send and email to step down.

Reviewer: ```suggestion
* If you are unable to participate send an email to step down.
```

>   We understand that short term committments can get in the way of sustainability initiatives – you are welcome to rejoin at any time.

Reviewer: ```suggestion
  We understand that short term commitments can get in the way of sustainability initiatives – you are welcome to rejoin at any time.
```

>   We will also retire inactive members as needed

Reviewer: ```suggestion
  We will also retire inactive members as needed.
```

  **Coordinated vulnerability disclosure**

  Disclosure policy:
  
  1. The reported vulnerability has been verified by working with the geoserver-security list
  2. GitHub `security advisory <https://github.com/geoserver/geoserver/security>`_ is used to reserve a CVE number
  3. A fix or documentation clarification is accepted and backported to both the "stable" and "maintenance" branches
  4. A fix is included for the "stable" and "maintenance" downloads (`released as scheduled <https://github.com/geoserver/geoserver/wiki/Release-Schedule>`__, or issued via emergency update)
  5. The CVE vulnerability is published with mitigation and patch instructions

  This represents a balance between transparency and participation that does not overwhelm participants. Those seeking greater visibility are encouraged to volunteer with the geoserver-security list; or work with one of the `commercial support providers <https://geoserver.org/support/>`__ who participate on behalf of their customers.

Working with vulnerability reports
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1. When working with a "triage" or "draft" vulnerability:

>    * Invite reported to participate, they are part of the team addressing the issue,

Reviewer: ```suggestion
   * Invite reporter to participate, they are part of the team addressing the issue,
```

>    * Give credit to the reporter any anyone else involved using the allocated fields

Reviewer: ```suggestion
   * Give credit to the reporter and anyone else involved using the allocated fields
```

   * Do not immediately request a CVE, and be sure to confirm the vulnerability (for example with a proof of concept) first.
   * A vulnerability flagged from a dependency scan does not automatically indicate that an exploit is available in GeoServer.
   * For popular concerns (like spring-framework upgrade) add each reporter to same advisory.
   

>    Be cautious that a security research may only be tasked with reporting the issue, and be unavailable

Reviewer: ```suggestion
   Be cautious that a security researcher may only be tasked with reporting the issue, and be unavailable
```

   once you have created a CVE.

2. Preparing report:

   * Package: Always report `org.geoserver.web:gs-web-app` as `geoserver.war` as a useful way to document that the `geoserver.war` includes other jars
   * Affected versions: It is difficult to communicate version ranges, due to limitations in CVE advistory processing, requiring multiple lines.
   
   .. figure:: img/cve-version-range.png
      
      Package guidance and version range

3. Work on providing a fix, mitigation instructions, or best-practice clarification for documentation.
   

>    * The use of github private repository associated with a advistory should be used with caution.

Reviewer: ```suggestion
   * The use of github private repository associated with an advisory should be used with caution.
```

   
     If making extensive code changes keep in mind that automations, including QA automations, are not available to be run as part of the pull-request review process.
     
     This may be appropriate when updating documentation / best practice information.
     
     They key advantage is several pull-requests can be managed at once, and merged at the same time as disclosure (see below).
   
   * If you wish to work on a public pull-request (to take advantage of workflow automations) take care that test-cases, commit messages, and documentation updates do not immediately reveal the vulnerability.
   

>    * Although not ideal it is possible to resolve some security issues by documenting a best-practice in the production consideration of the user guide.

Reviewer: ```suggestion
   * Although not ideal, it is possible to resolve some security issues by documenting a best-practice in the production consideration of the user guide.
```
